File: spring-data-requery-java/src/test/java/com/coupang/springframework/data/requery/domain/sample/AbstractUser.java
package com.coupang.springframework.data.requery.domain.sample;

import com.coupang.kotlinx.data.requery.conveters.ByteArrayBlobConverter;
import com.coupang.kotlinx.objectx.ToStringBuilder;
import com.coupang.springframework.data.requery.domain.AbstractPersistable;
import io.requery.*;
import lombok.Getter;
import lombok.Setter;
import org.jetbrains.annotations.NotNull;

import java.util.*;

/**
 * com.coupang.springframework.data.requery.domain.sample.AbstractUser
 *
 * @author debop
 * @since 18. 6. 14
 */
@Getter
@Setter
@Entity
@Table(name = "SD_User")
public abstract class AbstractUser extends AbstractPersistable<Integer> {

    private static final long serialVersionUID = -2977575626321229838L;

    @Key
    @Generated
    protected Integer id;

    protected String firstname;
    protected String lastname;
    protected int age;
    protected boolean active;

    protected Date createdAt;

    @Column(nullable = false, unique = true)
    protected String emailAddress;

    @JunctionTable
    @ManyToMany
    protected Set<AbstractUser> colleagues;

    @ManyToMany
    protected Set<AbstractRole> roles;

    @ManyToOne
    protected AbstractUser manager;

    @Embedded
    protected AbstractAddress address;

    @Convert(ByteArrayBlobConverter.class)
    protected byte[] binaryData;

    // NOTE: requery에서는 @ElementCollection은 지원하지 않습니다.
//    @ElementCollection
//    protected Set<String> attributes;

    protected Date dateOfBirth;

    public AbstractUser() {
        this(null, null, null);
    }

    public AbstractUser(String firstname, String lastname, String emailAddress, AbstractRole... roles) {
        this.firstname = firstname;
        this.lastname = lastname;
        this.emailAddress = emailAddress;
        this.active = true;
        this.roles = new HashSet<>(Arrays.asList(roles));

        this.createdAt = new Date();
    }

    @Override
    public int hashCode() {
        return Objects.hash(firstname, lastname, emailAddress);
    }

    @Transient
    @Override
    protected @NotNull ToStringBuilder buildStringHelper() {
        return super.buildStringHelper()
            .add("firstname", firstname)
            .add("lastname", lastname)
            .add("emailAddress", emailAddress);
    }
}

File: spring-data-requery-kotlin-test/src/main/kotlin/org/springframework/data/requery/kotlin/domain/sample/AbstractUser.kt
package org.springframework.data.requery.kotlin.domain.sample

import io.requery.Entity
import io.requery.Generated
import io.requery.Key
import io.requery.Table
import org.springframework.data.requery.kotlin.domain.AbstractPersistable
import org.springframework.data.requery.kotlin.domain.ToStringBuilder

/**
 * AbstractUser
 *
 * @author debop@coupang.com
 */
@Entity
@Table(name = "SD_User")
abstract class AbstractUser: AbstractPersistable<Int>() {

    @get:Key
    @get:Generated
    abstract override val id: Int

    abstract var firstname: String

    abstract var lastname: String

    abstract var email: String

    // NOTE: buildStringHelder 도 메소드이므로 @Transient 를 꼭 지정해줘야 한다.
    @io.requery.Transient
    override fun buildStringHelper(): ToStringBuilder {
        return super.buildStringHelper()
            .add("firstname", firstname)
            .add("lastname", lastname)
            .add("email", email)
    }
}